streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java

```diff
@@ -713,23 +713,18 @@ private boolean assignTasksToClients(final Set<String> allSourceTopics,
             allTasks, clientStates, numStandbyReplicas());
 
         final TaskAssignor taskAssignor;
-        if (highAvailabilityEnabled) {
-            if (lagComputationSuccessful) {
-                taskAssignor = new HighAvailabilityTaskAssignor(
-                    clientStates,
-                    allTasks,
-                    statefulTasks,
-                    assignmentConfigs);
-            } else {
-                log.info("Failed to fetch end offsets for changelogs, will return previous assignment to clients and "
-                             + "trigger another rebalance to retry.");
-                setAssignmentErrorCode(AssignorError.REBALANCE_NEEDED.code());
-                taskAssignor = new StickyTaskAssignor(clientStates, allTasks, statefulTasks, assignmentConfigs, true);
-            }
+        if (!lagComputationSuccessful) {
```

Comment:
This is a good thought. I think it mitigates the downside that we do still assign all the tasks when we fail to fetch lags, so it's not like we make no progress while waiting for the next rebalance. The "endless cycle" is a concern, but I'm not sure how it could happen in practice. I.e., what would make brokers consistently fail to report end offsets, but _not_ fail on any other APIs that Streams needs, especially since Streams needs to query the end-offset API during restoration anyway.

It seems like the failure would either be transient or permanent(ish).

If transient, then Streams will make progress during the probing.rebalance.interval, and succeed in balancing the assignment later. Even if we get further transient exceptions _during_ the sequence of HATA probing rebalances, the fact that we just return all tasks to their prior owners and that the HATA is stable mean that we just delay convergence by a single probing.rebalance.interval, not start all over again.

If permanent, then Streams will fail anyway _after_ the assignment completes, since it also  tends to query the end offsets immediately after getting the assignment. Even if it gets all prior tasks returned, which would make it skip the restoration phase, it seems implausible that we'd see a permanent failure on _only_ the end-offset API and Streams would happily be able to poll, commit, manage transactions, etc.

Our big alternative is just to immediately raise the exception, and leave it to KIP-572 to deal with the situation holistically. But I'm concerned that the impact of bombing out of assignment is greater than that of handling other failures during processing. It seems like an exception in assignment dooms the current Join/SyncGroup phase for everyone, which means that they have to wait for a timeout and then redo the rebalance. So KIP-572 can still recover gracefully, by reconstructing the consumer, but it can't help the extra downtime of waiting for the failed rebalance to time out and trying again.

Reply:
I agree with you on the principle. I think I'm just reaching a different conclusion.

I think we can leave aside the practical thought about the likelihood of total failure when the end-offset API fails permanently. That's really more of a supporting point that maybe this isn't a terrible idea.

The thing that makes me think that this is really preferable is that the task lags are an input to the TaskAssignor interface. It seems unreasonable for the StreamsPartitionAssignor would inspect the configured task assignor class and then decide just to pass in the ClientStates with one of the properties missing because it thinks it knows that particular assignor won't use it. Speaking of responsibility, it seems like it's not the responsibility of this class to reason about the implementation of each TaskAssignor. The purpose of an interface is that we don't have to worry about it, we just have to satisfy the interface. I think it would be just as strange, if not stranger to imagine coming back to the code base and trying to figure out why it's ok to pass a broken ClientState object just into the StickyTaskAssignor.

I actually can attest that that last point _is_ strange, having just spent a bunch of time in the integration tests, trying to figure out which combination of ClientState arguments are actually preconditions for `TaskAssignor#assign`, and why it was ok to mock some, but not all, of them, some of the time.

Reply:
I buy the argument that it's weird (aka bad) to supply an input that may be invalid just because we know that particular implementation doesn't use the input. Of course, we don't actually input the lags directly; we just happen to supplement the `ClientState` with the computed lags _in some cases_, and the HATA just happens to use that. It seems only slightly less weird to _not_ invoke something because we haven't added some supplemental information that isn't even required in the first place.

 What if we add a marker interface (warning: terrible placeholder name coming) called `UsingClientLagsAssignor` that only HATA implements to indicate that it needs this extra information on the `ClientState` to do its job. This will also be useful if we want to eventually make the task assignor fully user customizable; I expect someone who implemented a custom assignor that does not require task lags would be surprised (and annoyed) to find out that their custom assignor was skipped because we couldn't compute superfluous client data. 

Or, we could just move the lag computation into the HATA. But personally I'd rather leave it in the SPA for the above reason

Reply:
Probably a better alternative to the marker interface is just an abstract assignor class that adds the lag info to the `ClientState` which the HATA and potentially future custom assignors could implement

Reply:
Thanks @ableegoldman ,

I thought about the marker interface also, but didn't mention it because it seems like a bad sign if we have two implementations and two interfaces from the very start.

I think I'm getting a clue as to the contention when you mention "supplemental information". From my perspective, the TaskAssignor interface takes as input a "ClientState" for each instance in the cluster, which represents the current state of the cluster. One of the things it tells you is the lag for each task on the instance. How is this supplimental? It seems to be just one of the properties of the object. It actually happens to have a JavaDoc:

```java
    /**
     * Returns the total lag across all logged stores in the task. Equal to the end offset sum if this client
     * did not have any state for this task on disk.
     *
     * @return end offset sum - offset sum
     *          Task.LATEST_OFFSET if this was previously an active running task on this client
     */
    long lagFor(final TaskId task)
```

This is a private interface, so we can change this definition if it's not accurate. However, I'd be concerned about trying to program against the TaskAssignor interface if the caveat is that some of the arguments might be wrong or missing.

Reply:
Well, the `FallbackPriorTaskAssignor` also gets the same input but can't look at the lags. I figured if we're going to make a distinction between assignors that can check the lags and those that can't, StickyTaskAssignor should fall into the latter category.

Reply:
Hah, I got the last word!

Just kidding, fwiw I'm not trying to block this PR on the matter so it's fine by me if you merge as-is. Only wanted to make sure we're being fair to our assignor friends :P

Reply:
Hah! You know I can't let that happen :)

I think the fallback assignor is special in that, if we are unable to satisfy the TaskAssignor interface (because we couldn't compute lags), then at least we'd produce some kind of assignment. I.e., it's really just hitting the "abort" button on the whole assignment. In contrast to the internal, emergency-mode panic assignor, the StickyTaskAssignor is a regular, pluggable, assignor that people could use.

We'll have to revisit this topic anyway before making TaskAssignor a public API. Since it seems like both you and @cadonna view this change with suspicion, I'll add a special case for the StickyTaskAssignor, preserving the behavior before this PR.